pkg/runtime/adoption_reconciler.go

```diff
@@ -0,0 +1,468 @@
+	if err != nil {
+		return r.onError(ctx, desired, err)
+	}
+
+	rmo := described.RuntimeMetaObject()
+
+	// Use values from ReadOne output by default
+	targetMeta := &metav1.ObjectMeta{
+		Labels:          rmo.GetLabels(),
+		Annotations:     rmo.GetAnnotations(),
+		Finalizers:      rmo.GetFinalizers(),
+		OwnerReferences: rmo.GetOwnerReferences(),
+		GenerateName:    rmo.GetGenerateName(),
+	}
+
+	desiredMetadata := desired.Spec.Kubernetes.Metadata
+
+	// Attempt to use metadata values from the adopted resource target metadata
+	if desiredMetadata != nil {
+		if desiredMetadata.Name != "" {
+			targetMeta.SetName(desiredMetadata.Name)
+		}
+
+		if desiredMetadata.Namespace != "" {
+			targetMeta.SetNamespace(desiredMetadata.Namespace)
+		}
+
+		if len(desiredMetadata.Annotations) > 0 {
+			targetMeta.SetAnnotations(desiredMetadata.Annotations)
+		}
+
+		if len(desiredMetadata.Labels) > 0 {
+			targetMeta.SetLabels(desiredMetadata.Labels)
+		}
+
+		if len(desiredMetadata.OwnerReferences) > 0 {
+			targetMeta.SetOwnerReferences(desiredMetadata.OwnerReferences)
+		}
+
+		if desiredMetadata.GenerateName != "" {
+			targetMeta.SetGenerateName(desiredMetadata.GenerateName)
+		}
+	}
+
+	// If name and namespace not are specified, use the ones from the adopted
+	// resource directly.
+	if targetMeta.Name == "" {
+		targetMeta.SetName(desired.ObjectMeta.Name)
+	}
+
+	if targetMeta.Namespace == "" {
+		targetMeta.SetNamespace(desired.ObjectMeta.Namespace)
+	}
+
+	described.SetObjectMeta(*targetMeta)
+	targetDescriptor.MarkAdopted(described)
+
+	if err := r.kc.Create(ctx, described.RuntimeObject()); err != nil {
+		return r.onError(ctx, desired, err)
+	}
+
+	if err := r.markManaged(ctx, *desired); err != nil {
+		return r.onError(ctx, desired, err)
+	}
+
+	if err := r.onSuccess(ctx, desired); err != nil {
+		// Don't attempt to patch conditions again, directly return err
+		return err
+	}
+
+	return nil
+}
+
+// cleanup ensures that the supplied AWSResource's backing API resource is
+// destroyed along with all child dependent resources
+func (r *adoptionReconciler) cleanup(
+	ctx context.Context,
+	current ackv1alpha1.AdoptedResource,
+) error {
+	if err := r.markUnmanaged(ctx, current); err != nil {
+		return err
+	}
+	// Additional logic?
+	return nil
+}
+
+// getAdoptedResource returns an AdoptedResource representing the requested Kubernetes
+// namespaced object
+func (r *adoptionReconciler) getAdoptedResource(
+	ctx context.Context,
+	req ctrlrt.Request,
+) (*ackv1alpha1.AdoptedResource, error) {
+	ro := &ackv1alpha1.AdoptedResource{}
+	if err := r.kc.Get(ctx, req.NamespacedName, ro); err != nil {
+		return nil, err
+	}
+	return ro, nil
+}
+
+// onError will patch the adopted resource with the given error and return the
+// same error back
+func (r *adoptionReconciler) onError(
+	ctx context.Context,
+	res *ackv1alpha1.AdoptedResource,
+	err error,
+) error {
+	r.patchAdoptedCondition(ctx, res, err)
+	return err
```

Comment:
Error return value is not check here, did you mean `return r.patchAdoptedCondition(ctx, res, err)` ?

Reply:
Correct. This is meant to wrap an error, ensuring that it is implanted into the adopted resource before we raise it. It's a wrapper around `return err` in other methods.